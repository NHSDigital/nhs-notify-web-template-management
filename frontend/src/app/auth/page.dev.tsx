'use client';

import React, { Suspense } from 'react';
import { useSearchParams, redirect, RedirectType } from 'next/navigation';
import { Authenticator } from '@aws-amplify/ui-react';
import '@aws-amplify/ui-react/styles.css';
// this is an npm module, not node:path
// eslint-disable-next-line  unicorn/prefer-node-protocol
import path from 'path';
import { getBasePath } from '@utils/get-base-path';
import { NHSNotifyMain } from '@atoms/NHSNotifyMain/NHSNotifyMain';

const useRedirectPath = () => {
  const searchParams = useSearchParams();

  const requestRedirectPath = searchParams.get('redirect');

  const basePath = getBasePath();

  if (!requestRedirectPath) {
    return redirect('/', RedirectType.push);
  }

  let redirectPath = path.normalize(requestRedirectPath);

  if (redirectPath.startsWith(basePath)) {
    redirectPath = redirectPath.slice(basePath.length);
  }

  return redirectPath;
};

export const Redirect = () => {
  const redirectPath = useRedirectPath();

  return redirect(redirectPath, RedirectType.push);
};

const SignIn = () => {
  const redirectPath = useRedirectPath();

  redirect(`/auth/signin?redirect=${encodeURIComponent(redirectPath)}`);
};

export default function Page() {
  return (
    <NHSNotifyMain>
      <Suspense fallback={<p>Loading...</p>}>
        <Authenticator
          variation='default'
          hideSignUp
          formFields={{
            signIn: {
              username: {
                type: 'text',
              },
            },
          }}
        >
          <SignIn />
        </Authenticator>
      </Suspense>
    </NHSNotifyMain>
  );
}
