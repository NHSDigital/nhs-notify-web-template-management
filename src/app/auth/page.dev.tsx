/* eslint-disable @typescript-eslint/no-require-imports,unicorn/prefer-module,import/no-unresolved */

'use client';

import { Amplify } from 'aws-amplify';
import { Suspense } from 'react';
import { Authenticator, withAuthenticator } from '@aws-amplify/ui-react';
import '@aws-amplify/ui-react/styles.css';
import { Redirect } from '@molecules/Redirect/Redirect';

Amplify.configure(require('@/amplify_outputs.json'), { ssr: true });

const MockAuthPage = () =>
  withAuthenticator(Redirect, {
    variation: 'default',
    hideSignUp: true,
  })({});

const WrappedAuthPage = () => (
  <Authenticator.Provider>
    <Suspense>
      <MockAuthPage />
    </Suspense>
  </Authenticator.Provider>
);

export default WrappedAuthPage;
