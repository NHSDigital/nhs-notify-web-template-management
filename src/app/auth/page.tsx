'use client';

import { Amplify } from 'aws-amplify';
import { Suspense } from 'react';
import { withAuthenticator } from '@aws-amplify/ui-react';
import '@aws-amplify/ui-react/styles.css';
import { Redirect } from '@molecules/Redirect/Redirect';
import { getAmplifyOutputs } from '@utils/get-amplify-outputs';

Amplify.configure(getAmplifyOutputs(), { ssr: true });

const AuthPage = () =>
  withAuthenticator(Redirect, {
    variation: 'default',
    hideSignUp: true,
  })({});

const WrappedAuthPage = () => (
  <Suspense>
    <AuthPage />
  </Suspense>
);

export default WrappedAuthPage;
