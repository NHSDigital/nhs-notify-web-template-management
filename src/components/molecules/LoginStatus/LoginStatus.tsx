'use client';

import { Header } from 'nhsuk-react-components';
import React, { useEffect, useState } from 'react';
import { useAuthenticator } from '@aws-amplify/ui-react';
import { fetchAuthSession, signOut } from 'aws-amplify/auth';
import { JwtPayload } from 'aws-jwt-verify/jwt-model';

export default function LoginStatus() {

  const { authStatus } = useAuthenticator();
  const [idToken, setIdToken] = useState<JwtPayload | undefined>();
  useEffect(() => {
    (async () => {
      const session = await fetchAuthSession();
      setIdToken(session.tokens?.idToken?.payload);
    })().catch(console.error);
  }, [authStatus]);

  switch (authStatus) {
    case 'authenticated': {
      return [
        <Header.ServiceName key='serviceName'>
          {idToken?.email?.toString() || ''}
        </Header.ServiceName>,
        <Header.NavItem key='navItem' onClick={() => signOut()}>
          Sign out
        </Header.NavItem>,
      ];
    }
    case 'unauthenticated': {
      return (
        <Header.NavItem
          href={`https://chel5-auth-poc-rebased.d11o7gqmt8o0cx.amplifyapp.com/auth?redirect=${encodeURIComponent(location.href)}`} // eslint-disable-line no-restricted-globals
        >
          Sign in
        </Header.NavItem>
      );
    }
    default:
  }
}
