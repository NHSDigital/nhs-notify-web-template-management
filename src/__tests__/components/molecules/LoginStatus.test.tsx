import { mockDeep } from 'jest-mock-extended';
import { render, waitFor } from '@testing-library/react';
import { LoginStatus } from '@molecules/LoginStatus/LoginStatus';
import { Authenticator, useAuthenticator } from '@aws-amplify/ui-react';
import { fetchAuthSession } from '@aws-amplify/auth';

jest.mock('@aws-amplify/ui-react', () => ({
  ...jest.requireActual('@aws-amplify/ui-react'),
  useAuthenticator: jest.fn(),
}));
jest.mock('@aws-amplify/auth', () => ({
  ...jest.requireActual('@aws-amplify/auth'),
  fetchAuthSession: jest.fn(),
}));

test('LoginStatus - unauthenticated', async () => {
  jest.mocked(useAuthenticator).mockReturnValue(
    mockDeep<ReturnType<typeof useAuthenticator>>({
      authStatus: 'unauthenticated',
    })
  );
  jest.mocked(fetchAuthSession).mockResolvedValue({
    tokens: undefined,
  });

  const container = render(
    <Authenticator.Provider>
      <LoginStatus />
    </Authenticator.Provider>
  );

  await waitFor(() => container.getByText('Sign in'));
  expect(container.asFragment()).toMatchSnapshot();
});

test('LoginStatus - authenticated, no tokens', async () => {
  jest.mocked(useAuthenticator).mockReturnValue(
    mockDeep<ReturnType<typeof useAuthenticator>>({
      authStatus: 'authenticated',
    })
  );
  jest.mocked(fetchAuthSession).mockResolvedValue({
    tokens: undefined,
  });

  const container = render(
    <Authenticator.Provider>
      <LoginStatus />
    </Authenticator.Provider>
  );

  await waitFor(() => container.getByText('Sign out'));
  expect(container.asFragment()).toMatchSnapshot();
});

test('LoginStatus - authenticated', async () => {
  jest.mocked(useAuthenticator).mockReturnValue(
    mockDeep<ReturnType<typeof useAuthenticator>>({
      authStatus: 'authenticated',
    })
  );
  jest.mocked(fetchAuthSession).mockResolvedValue({
    tokens: {
      idToken: {
        payload: {
          email: 'test-user@nhs.net',
        },
      },
      accessToken: {
        payload: {},
      },
    },
  });

  const container = render(
    <Authenticator.Provider>
      <LoginStatus />
    </Authenticator.Provider>
  );

  await waitFor(() => container.getByText('Sign out'));
  expect(container.asFragment()).toMatchSnapshot();
});
