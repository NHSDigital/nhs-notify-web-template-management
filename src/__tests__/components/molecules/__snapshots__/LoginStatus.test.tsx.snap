// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginStatus - authenticated 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-header__transactional-service-name"
  >
    <a
      class="nhsuk-header__transactional-service-name--link"
    >
      test-user@nhs.net
    </a>
  </div>
  <li
    class="nhsuk-header__navigation-item"
  >
    <a
      class="nhsuk-header__navigation-link"
      href="/templates/auth/signout?redirect=%2F"
    >
      Sign out
    </a>
  </li>
</DocumentFragment>
`;

exports[`LoginStatus - authenticated, no tokens 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-header__transactional-service-name"
  >
    <a
      class="nhsuk-header__transactional-service-name--link"
    />
  </div>
  <li
    class="nhsuk-header__navigation-item"
  >
    <a
      class="nhsuk-header__navigation-link"
      href="/templates/auth/signout?redirect=%2F"
    >
      Sign out
    </a>
  </li>
</DocumentFragment>
`;

exports[`LoginStatus - unauthenticated 1`] = `
<DocumentFragment>
  <li
    class="nhsuk-header__navigation-item"
  >
    <a
      class="nhsuk-header__navigation-link"
      href="/templates/auth?redirect=%2F"
    >
      Sign in
    </a>
  </li>
</DocumentFragment>
`;
